examples/Mixtures/slot_based.py

```diff
@@ -0,0 +1,268 @@
+# with the same substance or just one holding the combined amounts of two slots
+# with the same slot ingredient. Thus, let us make sure that no slot contains a
+# duplicate label entry.
+
+no_duplicates_constraint = DiscreteNoLabelDuplicatesConstraint(
+    parameters=["Slot1_Label", "Slot2_Label", "Slot3_Label"]
+)
+
+# Next, we need to take care of permutation invariance. If our order of addition does
+# not matter, the result of exchanging slot 1 with slot 3 does not change the mixture,
+# i.e. the mixture slots are permutation invariant.
+
+# One complication arising for the permutation invariance in this case stems from the
+# fact that we not only have a label per slot, but also a numerical amount. Now if
+# this amount is zero, it actually does not matter what label the slot has (i.e.
+# what substance should be considered for that slot), because we are adding 0 of it to
+# the mixture anyway. In BayBE, we call this a "dependency", i.e. the slot labels
+# depend on the slot amounts and are only relevant if the amount fulfills some
+# condition (in this case "amount > 0"). The `DiscreteDependenciesConstraint` tells
+# the `DiscretePermutationInvarianceConstraint` about these dependencies so that they
+# are correctly included in the filtering process.
+
+perm_inv_constraint = DiscretePermutationInvarianceConstraint(
+    parameters=["Slot1_Label", "Slot2_Label", "Slot3_Label"],
+    dependencies=DiscreteDependenciesConstraint(
+        parameters=["Slot1_Amount", "Slot2_Amount", "Slot3_Amount"],
+        conditions=[
+            ThresholdCondition(threshold=0.0, operator=">"),
+            ThresholdCondition(threshold=0.0, operator=">"),
+            ThresholdCondition(threshold=0.0, operator=">"),
+        ],
+        affected_parameters=[["Slot1_Label"], ["Slot2_Label"], ["Slot3_Label"]],
+    ),
+)
+
+# Store all constraints.
+
+constraints = [perm_inv_constraint, sum_constraint, no_duplicates_constraint]
+
+# ```{admonition} Order of Addition
+# :class: note
+# Whether you need to impose the constraints for removing duplicates and
+# permutation invariance depends on your use case. If the order of addition is relevant
+# to your mixture, there is no permutation invariance and one could argue that
+# duplicates should also be allowed if subsequent steps can add the same substance.
+# ```
+
+### Campaign Setup
+
+searchspace = SearchSpace.from_product(parameters=parameters, constraints=constraints)
+objective = NumericalTarget(name="Target_1", mode="MAX").to_objective()
+campaign = Campaign(searchspace=searchspace, objective=objective)
+
+### Verification of Constraints
+
+# Now let us take a look at some recommendations for this campaign and check whether
+# the constraints we imposed are indeed adhered to.
+
+N_ITERATIONS = 2 if SMOKE_TEST else 3
+for kIter in range(N_ITERATIONS):
+    print(f"\n#### ITERATION {kIter+1} ####")
+
+    print("## ASSERTS ##")
```

Comment:
Technically, we do not use any asserts here. Can't we just change that and also include them?

Reply:
i wouldnt like asserts here as it would havemuch unnecessary code, the point it to print something visual

should `assert` be renamed? How important is this?

Reply:
Hi @Scienfitz, perhaps wait with the refactoring of that until I have completed the review to save some double work

Reply:
pushed the small changes, did not touch this section yet

Reply:
Ok, so here my thoughts. I have several problems with the lower part:
* What it does is completely unrelated to any `recommend` iterations! What is the point of those
* In fact, it is also unrelated to the campaign itself! At this point, I would like to repeat what I've suggested already a few times: There is no need that all our examples always end up with i) the creation of a campaign plus ii) running iterations on it. The purpose of this example is to illustrate the search space creation, and that is where it should logically end.
* The `No. of unique x-solvent entries` parts technically have nothing to do with the constraints. If we want to keep it, it should be moved to a separate block because:
* The entire section is way too much code/content for my taste
* The code does not actually `assert` (i.e. in the python sense) things. But adding actual asserts would be useful for additional testing.

So my recommendation:
* Let us get rid of the campaing/iteration section --> No more SMOKE_TEST needed
* We crank down the numbers such that we can actually print the search space and people can look at it
* We perform the necessary asserts as real `assert`
* If we want to keep it, we show the statistics in a separate section

What do you think?

Reply:
in order of your bullet points (suggest using enum instead of bullets in the future for easy reference)

1. OK agree about recommend statements and campaign, they can go and this can only do stuff based on the discrete searchspace. Keep in mind though that this will make it inconsistent with the traditional example - there we need recommendations to demonstrate the constraint effect (contrary to the discrete ansatz in here)
2. see 1. not relevant in this PR, but please consider the continuous case, there what you say is not so easily true, how to demonstrate the effect of settings and constraints without producing recommendations?
3. You must elaborate on that, because these asserts are the most important ones demosntrating the combined effect of removing permutaiton invariant + duplicate labels as well as incorporating the dependency between labels and amounts
4. Guess thats fixed by compressing the assert part which seems bloated and you dont seem to like. But then also, the slot-based mixture modelling is easily the most complex example we have, not sure how reasonable it is to want this to be short. I definitely prefer a complete example especially if its about a complex topic
5. The word `assert` is just a word, we can easily call that something else, the point of that section is to demonstrate the effect of constraints on the searchspace. We have similar tests in the proper tests already, so actual `assert`s imo are not needed, especially if this example is already too long in your view

Reply:
Hey @Scienfitz, I've just pushed a version that shows how I'd do it. Let me know what you like/dislike. BTW: have added an explanation for the theoretical number – now I know why you had a hard time. Combinatorics, gotta love it :D

Reply:
Hey, no bad word about combinatorics here 😠  That stuff is super cool. I had a full course on it, you don't even want to know what and how we counted 🔢 
Content wise: I only wanted to change the word `assert` (or include actual `assert` statements) since this is a pre-occupied word in python. I did not want to start this huge discussion 😆 
After having a brief look at both implementations, I do have a preferred version but will let the two of you discuss first :)

Reply:
Alright, then I'm happy. I just really wonder how you have derived it otherwise 🤔 Since you didn't have a general formula, it seemed like you came up with individual solutions for the different cases? Anyway, no longer important now but in case you ever gonna remember again: let me know 🙃